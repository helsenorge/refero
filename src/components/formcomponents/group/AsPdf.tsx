import { ReactNode } from 'react';

type Props = {
  pdf: boolean;
  children: ReactNode;
};

const AsPdf = ({ pdf, children }: Props): JSX.Element => {
  return <>{pdf ? <div>{children}</div> : children}</>;
};
export default AsPdf;
