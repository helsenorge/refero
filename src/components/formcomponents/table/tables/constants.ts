export const DATEFORMATS = {
  DATE: 'DD.MM.YYYY',
  DATETIME: 'DD MM YYYY HH:mm',
  TIME: 'HH:mm',
};
