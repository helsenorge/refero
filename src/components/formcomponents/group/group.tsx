import * as React from 'react';

import DOMPurify from 'dompurify';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';

import { QuestionnaireItem, QuestionnaireResponseItemAnswer, QuestionnaireResponseItem, Questionnaire } from '../../../types/fhir';

import AnchorLink from '@helsenorge/designsystem-react/components/AnchorLink';

import { NewValueAction } from '../../../actions/newValue';
import { RenderContextType } from '../../../constants/renderContextType';
import { GlobalState } from '../../../reducers';
import { getGroupItemControl } from '../../../util/group-item-control';
import { renderPrefix, getText, getId } from '../../../util/index';
import { mapStateToProps, mergeProps, mapDispatchToProps } from '../../../util/map-props';
import { RenderContext } from '../../../util/renderContext';
import { Resources } from '../../../util/resources';
import { Path } from '../../../util/refero-core';
import withCommonFunctions from '../../with-common-functions';

export interface Props {
  item: QuestionnaireItem;
  questionnaire?: Questionnaire;
  answer: QuestionnaireResponseItemAnswer;
  responseItem: QuestionnaireResponseItem;
  dispatch?: ThunkDispatch<GlobalState, void, NewValueAction>;
  path: Array<Path>;
  pdf?: boolean;
  includeSkipLink?: boolean;
  className?: string;
  resources?: Resources;
  headerTag?: number;
  attachmentErrorMessage?: string;
  renderDeleteButton: (className?: string) => JSX.Element | undefined;
  renderChildrenItems: (renderContext: RenderContext) => Array<JSX.Element> | undefined;
  repeatButton: JSX.Element;
  id?: string;
  renderContext: RenderContext;
  renderHelpButton: () => JSX.Element;
  renderHelpElement: () => JSX.Element;
  isHelpOpen?: boolean;
  onRenderMarkdown?: (item: QuestionnaireItem, markdown: string) => string;
}
interface State {
  counter?: number;
}
export class Group extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
  }

  shouldComponentUpdate(nextProps: Props): boolean {
    const responseItemHasChanged = this.props.responseItem !== nextProps.responseItem;
    const helpItemHasChanged = this.props.isHelpOpen !== nextProps.isHelpOpen;
    const repeatButtonHasChanged = this.props.repeatButton !== nextProps.repeatButton;
    const attachmentErrorMessageHasChanged = this.props.attachmentErrorMessage !== nextProps.attachmentErrorMessage;
    const resourcesHasChanged = JSON.stringify(this.props.resources) !== JSON.stringify(nextProps.resources);

    const repeats = this.props.item.repeats ?? false;

    return (
      responseItemHasChanged ||
      helpItemHasChanged ||
      repeatButtonHasChanged ||
      attachmentErrorMessageHasChanged ||
      resourcesHasChanged ||
      repeats
    );
  }

  renderAllItems = (): JSX.Element => {
    const { renderContext } = this.props;
    const localRenderContextType = this.getLocalRenderContextType();

    if (localRenderContextType) {
      switch (localRenderContextType) {
        case RenderContextType.Grid:
          return this.renderContextTypeGrid();
      }
    }

    switch (renderContext.RenderContextType) {
      case RenderContextType.Grid:
        return this.isDirectChildOfRenderContextOwner() ? this.renderContextTypeGridRow() : this.renderGroup();
      default:
        return this.renderGroup();
    }
  };

  isDirectChildOfRenderContextOwner = (): boolean => {
    const { path, item, renderContext } = this.props;

    const myIndex = path.findIndex(p => p.linkId === item.linkId);
    if (myIndex > 0) {
      const directParentLinkId = path[myIndex - 1].linkId;
      return directParentLinkId === renderContext.Owner;
    }

    return false;
  };

  renderContextTypeGridRow = (): JSX.Element => {
    const { renderContext, item } = this.props;

    renderContext.RenderChildren = (
      childItems: QuestionnaireItem[],
      itemRenderer: (item: QuestionnaireItem, renderContext: RenderContext) => Array<JSX.Element | undefined>
    ): JSX.Element[] => {
      const renderedChildItems = [];
      let counter = 1;
      for (const column of renderContext.Columns) {
        const childItem = childItems.find(item => item.text === column);

        if (childItem) {
          renderedChildItems.push(
            <td key={counter} className="page_refero__grid--cell">
              {itemRenderer(childItem, renderContext)}
            </td>
          );
        } else {
          renderedChildItems.push(<td key={counter} className="page_refero__grid--cell page_refero__grid--cell-empty">{` `}</td>);
        }

        counter++;
      }

      return renderedChildItems;
    };

    return (
      <tr key={item.linkId} className="page_refero__grid--row">
        <td className="page_refero__grid--cell page_refero__grid--cell-first">{this.renderGroupHeader()}</td>
        {this.props.renderChildrenItems(renderContext)}
      </tr>
    );
  };

  renderContextTypeGrid = (): JSX.Element => {
    const { item } = this.props;

    const columns = this.getColumns();
    const headers = columns.map(c => <th key={item.linkId + '-' + c}>{c}</th>);
    headers.unshift(<th key={item.linkId + 'X'}>{item.text ? item.text : ''}</th>);

    const newRenderContext = new RenderContext(RenderContextType.Grid, item.linkId, columns);
    return (
      <React.Fragment>
        <table id={getId(this.props.id)} className="page_refero__grid">
          <thead>
            <tr>{headers}</tr>
          </thead>
          <tbody>{this.props.renderChildrenItems(newRenderContext)}</tbody>
        </table>
        {this.props.renderDeleteButton('page_refero__deletebutton--margin-top')}
        {this.props.repeatButton}
      </React.Fragment>
    );
  };

  renderGroup = (): JSX.Element => {
    return (
      <section id={getId(this.props.id)} data-sectionname={this.getHeaderText()}>
        {this.renderGroupHeader()}
        {this.props.renderHelpElement()}
        <div id={`${getId(this.props.id)}-navanchor`} className={this.getClassNames()}>
          {this.props.renderChildrenItems(new RenderContext())}
        </div>
        {this.props.includeSkipLink && this.props.path.length === 1 && (
          <AnchorLink className="page_refero__skiplink" href="#navigator-button">
            {this.props.resources?.skipLinkText}
          </AnchorLink>
        )}
        {this.props.renderDeleteButton('page_refero__deletebutton--margin-top')}
        {this.props.repeatButton}
      </section>
    );
  };

  getColumns = (): Array<string> => {
    const item = this.props.item;
    const seenColumns = {};
    const columns: Array<string> = [];
    if (!item.item || item.item.length === 0) return columns;
    for (const group of item.item) {
      if (group.item && group.item.length > 0) {
        for (const cell of group.item) {
          const key = cell.text || '';
          if (key in seenColumns) continue;

          columns.push(key);
          seenColumns[key] = 1;
        }
      }
    }

    return columns;
  };

  getLocalRenderContextType = (): RenderContextType => {
    const coding = getGroupItemControl(this.props.item);
    if (coding.length > 0) {
      switch (coding[0].code) {
        case 'grid':
          return RenderContextType.Grid;
      }
    }
    return RenderContextType.None;
  };

  getClassNames = (): string => {
    const classNames = ['page_refero__component', 'page_refero__component_group'];
    const coding = getGroupItemControl(this.props.item);
    if (coding.length > 0) {
      classNames.push('page_refero__itemControl_' + coding[0].code);
    }

    return classNames.join(' ');
  };

  getComponentToValidate = (): undefined => {
    return undefined;
  };

  getHeaderText = (): string => {
    return (
      renderPrefix(this.props.item) +
      ' ' +
      getText(this.props.item, this.props.onRenderMarkdown, this.props.questionnaire, this.props.resources)
    );
  };

  renderGroupHeader = (): JSX.Element | null => {
    if (!getText(this.props.item, this.props.onRenderMarkdown)) {
      return null;
    }

    const HeaderTag = `h${this.props.headerTag}` as 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
    const headerText = DOMPurify.sanitize(this.getHeaderText(), {
      RETURN_TRUSTED_TYPE: true,
      ADD_ATTR: ['target'],
    }) as unknown as string;
    return (
      <React.Fragment>
        <HeaderTag className={'page_refero__heading'} dangerouslySetInnerHTML={{ __html: headerText }} />
        {this.props.renderHelpButton()}
      </React.Fragment>
    );
  };

  render(): JSX.Element | null {
    const { pdf } = this.props;

    if (pdf) {
      return <div>{this.renderAllItems()}</div>;
    }

    return this.renderAllItems();
  }
}
const withCommonFunctionsComponent = withCommonFunctions(Group);
const connectedComponent = connect(mapStateToProps, mapDispatchToProps, mergeProps)(withCommonFunctionsComponent);
export default connectedComponent;
