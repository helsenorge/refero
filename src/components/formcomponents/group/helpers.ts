import { Questionnaire, QuestionnaireItem } from 'fhir/r4';

import { RenderContextType } from '../../../constants/renderContextType';
import { getText, renderPrefix } from '../../../util';
import { getGroupItemControl } from '../../../util/group-item-control';
import { Path } from '../../../util/refero-core';
import { RenderContext } from '../../../util/renderContext';
import { Resources } from '../../../util/resources';

export const getColumns = (item: QuestionnaireItem): Array<string> => {
  const seenColumns = {};
  const columns: Array<string> = [];
  if (!item.item || item.item.length === 0) return columns;
  for (const group of item.item) {
    if (group.item && group.item.length > 0) {
      for (const cell of group.item) {
        const key = cell.text || '';
        if (key in seenColumns) continue;

        columns.push(key);
        seenColumns[key] = 1;
      }
    }
  }

  return columns;
};

export const getLocalRenderContextType = (item: QuestionnaireItem): RenderContextType => {
  const coding = getGroupItemControl(item);
  if (coding.length > 0) {
    switch (coding[0].code) {
      case 'grid':
        return RenderContextType.Grid;
    }
  }
  return RenderContextType.None;
};

export const isDirectChildOfRenderContextOwner = (path: Path[], item: QuestionnaireItem, renderContext: RenderContext): boolean => {
  const myIndex = path.findIndex(p => p.linkId === item.linkId);
  if (myIndex > 0) {
    const directParentLinkId = path[myIndex - 1].linkId;
    return directParentLinkId === renderContext.Owner;
  }

  return false;
};

export const getClassNames = (item: QuestionnaireItem): string => {
  const classNames = ['page_refero__component', 'page_refero__component_group'];
  const coding = getGroupItemControl(item);
  if (coding.length > 0) {
    classNames.push('page_refero__itemControl_' + coding[0].code);
  }

  return classNames.join(' ');
};

export const getHeaderText = (
  item: QuestionnaireItem,
  questionnaire?: Questionnaire,
  resources?: Resources,
  onRenderMarkdown?: ((item: QuestionnaireItem, markdown: string) => string) | undefined
): string => {
  return renderPrefix(item) + ' ' + getText(item, onRenderMarkdown, questionnaire, resources);
};
