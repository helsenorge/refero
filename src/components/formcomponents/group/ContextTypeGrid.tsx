import { getId } from '@/util';
import RenderDeleteButton from '../repeat/RenderDeleteButton';
import RenderRepeatButton from '../repeat/RenderRepeatButton';
import { RenderContext } from '@/util/renderContext';
import { getColumns } from './helpers';

import { RenderContextType } from '@/constants/renderContextType';
import { QuestionnaireItem, QuestionnaireResponseItem } from 'fhir/r4';
import GenerateQuestionnaireComponents, {
  QuestionnaireComponentItemProps,
} from '@/components/createQuestionnaire/GenerateQuestionnaireComponents';
import { useExternalRenderContext } from '@/context/externalRenderContext';

type ContextTypeGridProps = QuestionnaireComponentItemProps & {
  item?: QuestionnaireItem;
  responseItem?: QuestionnaireResponseItem;
};
type Props = {
  item?: QuestionnaireItem;
  columns: string[];
};
const RenderHeaders = ({ item, columns }: Props): JSX.Element => {
  const headers = columns.map((c, i) => <th key={`${item?.linkId}-${c}-${i}`}>{c}</th>);
  headers.unshift(<th key={`${item?.linkId}-X`}>{item?.text ? item.text : ''}</th>);
  return <>{headers}</>;
};

const ContextTypeGrid = (props: ContextTypeGridProps): JSX.Element => {
  const { item, index, path, id, responseItem } = props;
  const columns = getColumns(item);
  const { resources } = useExternalRenderContext();
  return (
    <>
      <table id={getId(id)} className="page_refero__grid">
        <thead>
          <tr>
            <RenderHeaders item={item} columns={columns} />
          </tr>
        </thead>
        <tbody>
          <GenerateQuestionnaireComponents
            {...props}
            items={item?.item}
            renderContext={new RenderContext(RenderContextType.Grid, item?.linkId, columns)}
          />
        </tbody>
      </table>
      <RenderDeleteButton
        item={item}
        path={path}
        index={index}
        responseItem={responseItem}
        className="page_refero__deletebutton--margin-top"
      />
      <RenderRepeatButton item={item} index={index} path={path?.slice(0, -1)} resources={resources} responseItem={responseItem} />
    </>
  );
};

export default ContextTypeGrid;
