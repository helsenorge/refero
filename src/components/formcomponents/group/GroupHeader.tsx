import SafeText from '@/components/referoLabel/SafeText';
import RenderHelpButton from '../help-button/RenderHelpButton';
import { getHeaderText } from './helpers';
import { getText } from '@/util';
import { useExternalRenderContext } from '@/context/externalRenderContext';
import { Questionnaire, QuestionnaireItem } from 'fhir/r4';
import { Resources } from '@/util/resources';
import { Dispatch } from 'react';
import { useSelector } from 'react-redux';
import { getFormDefinition } from '@/reducers/form';
import { GlobalState } from '@/reducers';
import RenderHelpElement from '../help-button/RenderHelpElement';

type GroupHeaderProps = {
  item?: QuestionnaireItem;
  questionnaire?: Questionnaire | null;
  resources?: Resources;
  headerTag?: number;
  isHelpVisible: boolean;
  setIsHelpVisible: Dispatch<React.SetStateAction<boolean>>;
};
const GroupHeader = ({ item, resources, headerTag, isHelpVisible, setIsHelpVisible }: GroupHeaderProps): JSX.Element | null => {
  const { onRenderMarkdown } = useExternalRenderContext();
  const formDefinition = useSelector((state: GlobalState) => getFormDefinition(state));
  const questionnaire = formDefinition?.Content;
  if (!getText(item, onRenderMarkdown)) {
    return null;
  }

  const HeaderTag = `h${headerTag ?? 2}` as 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
  const headerText = getHeaderText(item, questionnaire, resources, onRenderMarkdown);
  return (
    <>
      <SafeText as={HeaderTag} text={headerText} className={'page_refero__heading'} />
      <RenderHelpButton isHelpVisible={isHelpVisible} item={item} setIsHelpVisible={setIsHelpVisible} padding={true} />
      <RenderHelpElement isHelpVisible={isHelpVisible} item={item} />
    </>
  );
};
export default GroupHeader;
