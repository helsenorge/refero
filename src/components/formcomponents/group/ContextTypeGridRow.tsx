import { QuestionnaireItem } from 'fhir/r4';
import GroupHeader from './GroupHeader';
import { RenderContext } from '@/util/renderContext';
import { Dispatch } from 'react';
import { RenderChildrenItems, RenderItemProps } from '../renderChildren/RenderChildrenItems';

type ContextTypeGridRowProps = RenderItemProps & {
  isHelpVisible: boolean;
  setIsHelpVisible: Dispatch<React.SetStateAction<boolean>>;
};
const ContextTypeGridRow = (props: ContextTypeGridRowProps): JSX.Element => {
  const { isHelpVisible, setIsHelpVisible, item, resources, renderContext, headerTag, questionnaire } = props;

  renderContext.RenderChildren = (
    childItems: QuestionnaireItem[],
    itemRenderer: (item: QuestionnaireItem, renderContext: RenderContext) => JSX.Element | null
  ): JSX.Element[] => {
    const renderedChildItems = [];
    let counter = 1;
    for (const column of renderContext.Columns) {
      const childItem = childItems.find(item => item.text === column);

      if (childItem) {
        renderedChildItems.push(
          <td key={counter} className="page_refero__grid--cell">
            {itemRenderer(childItem, renderContext)}
          </td>
        );
      } else {
        renderedChildItems.push(<td key={counter} className="page_refero__grid--cell page_refero__grid--cell-empty">{` `}</td>);
      }

      counter++;
    }

    return renderedChildItems;
  };

  return (
    <tr key={item.linkId} className="page_refero__grid--row">
      <td className="page_refero__grid--cell page_refero__grid--cell-first">
        <GroupHeader
          headerTag={headerTag}
          isHelpVisible={isHelpVisible}
          setIsHelpVisible={setIsHelpVisible}
          item={item}
          questionnaire={questionnaire}
          resources={resources}
        />
      </td>
      {<RenderChildrenItems otherProps={props} />}
    </tr>
  );
};
export default ContextTypeGridRow;
