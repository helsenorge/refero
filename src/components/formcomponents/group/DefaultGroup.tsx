import { getId } from '@/util';
import RenderHelpElement from '../help-button/RenderHelpElement';
import GroupHeader from './GroupHeader';
import { getClassNames, getHeaderText } from './helpers';
import AnchorLink from '@helsenorge/designsystem-react/components/AnchorLink';
import RenderDeleteButton from '../repeat/RenderDeleteButton';
import styles from './group.module.css';
import { Dispatch } from 'react';

import { useExternalRenderContext } from '@/context/externalRenderContext';
import RenderRepeatButton from '../repeat/RenderRepeatButton';
import { QuestionnaireComponentItemProps } from '@/components/createQuestionnaire/GenerateQuestionnaireComponents';
import { useSelector } from 'react-redux';
import { GlobalState } from '@/reducers';
import { getFormDefinition } from '@/reducers/form';

type DefaultGroup = QuestionnaireComponentItemProps & {
  isHelpVisible: boolean;
  setIsHelpVisible: Dispatch<React.SetStateAction<boolean>>;
};
const DefaultGroup = ({ isHelpVisible, setIsHelpVisible, children, ...rest }: DefaultGroup): JSX.Element => {
  const { item, resources, headerTag, includeSkipLink, path, responseItem, index, id, responseItems } = rest;
  const { onRenderMarkdown } = useExternalRenderContext();
  const formDefinition = useSelector((state: GlobalState) => getFormDefinition(state));
  const questionnaire = formDefinition?.Content;

  return (
    <section id={getId(id)} data-sectionname={getHeaderText(item, questionnaire, resources, onRenderMarkdown)}>
      <GroupHeader
        headerTag={headerTag}
        isHelpVisible={isHelpVisible}
        setIsHelpVisible={setIsHelpVisible}
        item={item}
        resources={resources}
      />
      <RenderHelpElement item={item} isHelpVisible={isHelpVisible} />
      {includeSkipLink && path?.length === 1 && (
        <AnchorLink className={`${styles.page_refero__skiplink} page_refero__skiplink`} href="#navigator-button">
          {resources?.skipLinkText}
        </AnchorLink>
      )}
      <div id={`${getId(id)}-navanchor`} className={getClassNames(item)}>
        {children}
      </div>

      <RenderDeleteButton
        item={item}
        path={path}
        index={index}
        responseItem={responseItem}
        className="page_refero__deletebutton--margin-top"
      />
      <RenderRepeatButton
        item={item}
        index={index}
        path={path?.slice(0, -1)}
        resources={resources}
        responseItem={responseItem}
        responseItems={responseItems}
      />
    </section>
  );
};
export default DefaultGroup;
