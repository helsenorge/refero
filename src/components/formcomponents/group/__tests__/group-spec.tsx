import * as React from 'react';
import { shallow } from 'enzyme';

import '../../../../util/defineFetch';
import { Group } from '../group';
import { QuestionnaireItem, QuestionnaireResponseItemAnswer, QuestionnaireResponseItem } from '../../../../types/fhir';
import { RenderContextType } from '../../../../constants/renderContextType';
import { RenderContext } from '../../../../util/renderContext';

describe('group', () => {
  it('should render correct tag', () => {
    const item: QuestionnaireItem = {
      id: '2',
      linkId: '2.1',
      repeats: false,
      type: 'group',
      text: 'Overskrift',
    };
    const answer: QuestionnaireResponseItemAnswer = {} as QuestionnaireResponseItemAnswer;
    const group = shallow(
      <Group
        item={item}
        answer={answer}
        path={[]}
        headerTag={3}
        renderChildrenItems={() => {
          return undefined;
        }}
        repeatButton={<div />}
        renderDeleteButton={() => {
          return undefined;
        }}
        id="item_2"
        renderHelpButton={() => <React.Fragment />}
        renderHelpElement={() => <React.Fragment />}
        renderContext={new RenderContext(RenderContextType.None)}
        responseItem={{} as QuestionnaireResponseItem}
      />
    );
    expect(group.find('h3').html()).toMatchSnapshot();
  });
});
