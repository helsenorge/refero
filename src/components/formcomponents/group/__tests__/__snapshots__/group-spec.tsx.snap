// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`group should render correct tag 1`] = `"<h3 class=\\"page_refero__heading\\"> Overskrift</h3>"`;
